File: backend/app/models/savings.py
from datetime import datetime, date
from decimal import Decimal
import enum
from sqlalchemy import (
    Column,
    Integer,
    String,
    Numeric,
    DateTime,
    Date,
    Enum,
    ForeignKey,
    Text,
)
from sqlalchemy.orm import relationship

from app.core.database import Base


class SavingsGroupStatus(str, enum.Enum):
    ACTIVE = "ACTIVE"
    CLOSED = "CLOSED"


class SavingsMemberStatus(str, enum.Enum):
    ACTIVE = "ACTIVE"
    INACTIVE = "INACTIVE"


class SavingsPaymentMethod(str, enum.Enum):
    CASH = "CASH"
    MPESA = "MPESA"
    TRANSFER = "TRANSFER"


class SavingsLoanStatus(str, enum.Enum):
    ACTIVE = "ACTIVE"
    REPAID = "REPAID"
    DEFAULTED = "DEFAULTED"


class SavingsGroup(Base):
    __tablename__ = "savings_groups"

    id = Column(Integer, primary_key=True, index=True)
    company_id = Column(Integer, ForeignKey("companies.id", ondelete="CASCADE"), nullable=False, index=True)
    name = Column(String(255), nullable=False)
    interest_rate = Column(Numeric(5, 2), default=Decimal("10.00"), nullable=False)  # % mensal
    cycle_months = Column(Integer, default=12, nullable=False)
    status = Column(Enum(SavingsGroupStatus), default=SavingsGroupStatus.ACTIVE, nullable=False)
    start_date = Column(Date, default=date.today, nullable=False)
    end_date = Column(Date, nullable=True)
    admin_user_id = Column(Integer, ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    members = relationship("SavingsMember", back_populates="group", cascade="all, delete-orphan")
    deposits = relationship("SavingsDeposit", back_populates="group", cascade="all, delete-orphan")
    loans = relationship("SavingsLoan", back_populates="group", cascade="all, delete-orphan")


class SavingsMember(Base):
    __tablename__ = "savings_members"

    id = Column(Integer, primary_key=True, index=True)
    group_id = Column(Integer, ForeignKey("savings_groups.id", ondelete="CASCADE"), nullable=False, index=True)
    name = Column(String(255), nullable=False)
    phone = Column(String(50), nullable=False)
    total_deposited = Column(Numeric(12, 2), default=Decimal("0.00"), nullable=False)
    total_borrowed = Column(Numeric(12, 2), default=Decimal("0.00"), nullable=False)
    total_repaid = Column(Numeric(12, 2), default=Decimal("0.00"), nullable=False)
    status = Column(Enum(SavingsMemberStatus), default=SavingsMemberStatus.ACTIVE, nullable=False)
    joined_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    group = relationship("SavingsGroup", back_populates="members")
    deposits = relationship("SavingsDeposit", back_populates="member", cascade="all, delete-orphan")
    loans = relationship("SavingsLoan", back_populates="member", cascade="all, delete-orphan")


class SavingsDeposit(Base):
    __tablename__ = "savings_deposits"

    id = Column(Integer, primary_key=True, index=True)
    group_id = Column(Integer, ForeignKey("savings_groups.id", ondelete="CASCADE"), nullable=False, index=True)
    member_id = Column(Integer, ForeignKey("savings_members.id", ondelete="CASCADE"), nullable=False, index=True)
    amount = Column(Numeric(12, 2), nullable=False)
    deposited_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    payment_method = Column(Enum(SavingsPaymentMethod), default=SavingsPaymentMethod.CASH, nullable=False)
    notes = Column(Text, nullable=True)

    group = relationship("SavingsGroup", back_populates="deposits")
    member = relationship("SavingsMember", back_populates="deposits")


class SavingsLoan(Base):
    __tablename__ = "savings_loans"

    id = Column(Integer, primary_key=True, index=True)
    group_id = Column(Integer, ForeignKey("savings_groups.id", ondelete="CASCADE"), nullable=False, index=True)
    member_id = Column(Integer, ForeignKey("savings_members.id", ondelete="CASCADE"), nullable=False, index=True)
    amount = Column(Numeric(12, 2), nullable=False)
    interest_rate = Column(Numeric(5, 2), nullable=False)
    monthly_payment = Column(Numeric(12, 2), default=Decimal("0.00"), nullable=False)
    total_repayable = Column(Numeric(12, 2), nullable=False)
    amount_repaid = Column(Numeric(12, 2), default=Decimal("0.00"), nullable=False)
    disbursed_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    due_date = Column(Date, nullable=False)
    status = Column(Enum(SavingsLoanStatus), default=SavingsLoanStatus.ACTIVE, nullable=False)

    group = relationship("SavingsGroup", back_populates="loans")
    member = relationship("SavingsMember", back_populates="loans")
    repayments = relationship("SavingsRepayment", back_populates="loan", cascade="all, delete-orphan")


class SavingsRepayment(Base):
    __tablename__ = "savings_repayments"

    id = Column(Integer, primary_key=True, index=True)
    loan_id = Column(Integer, ForeignKey("savings_loans.id", ondelete="CASCADE"), nullable=False, index=True)
    amount = Column(Numeric(12, 2), nullable=False)
    principal_portion = Column(Numeric(12, 2), default=Decimal("0.00"), nullable=False)
    interest_portion = Column(Numeric(12, 2), default=Decimal("0.00"), nullable=False)
    repaid_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    payment_method = Column(Enum(SavingsPaymentMethod), default=SavingsPaymentMethod.CASH, nullable=False)

    loan = relationship("SavingsLoan", back_populates="repayments")

File: frontend/src/types/savings.ts
export type SavingsGroupStatus = 'ACTIVE' | 'CLOSED';
export type SavingsMemberStatus = 'ACTIVE' | 'INACTIVE';
export type SavingsPaymentMethod = 'CASH' | 'MPESA' | 'TRANSFER';
export type SavingsLoanStatus = 'ACTIVE' | 'REPAID' | 'DEFAULTED';

export interface SavingsMember {
  id: number;
  member_id?: number;
  group_id?: number;
  name: string;
  phone: string;
  total_deposited: number;
  total_borrowed: number;
  total_repaid: number;
  status: SavingsMemberStatus;
  joined_at?: string;
  share_pct?: number;
  share_value?: number;
}

export interface SavingsGroup {
  id: number;
  name: string;
  interest_rate: number;
  cycle_months: number;
  status: SavingsGroupStatus;
  start_date: string;
  end_date?: string | null;
  total_members?: number;
  created_at?: string;
  members?: SavingsMember[];
}

export interface SavingsDeposit {
  id: number;
  group_id: number;
  member_id: number;
  amount: number;
  payment_method: SavingsPaymentMethod;
  notes?: string | null;
  deposited_at: string;
}

export interface SavingsLoan {
  id: number;
  group_id?: number;
  member_id: number;
  member_name?: string;
  amount: number;
  interest_rate: number;
  monthly_payment: number;
  total_repayable: number;
  amount_repaid: number;
  remaining?: number;
  due_date: string;
  status: SavingsLoanStatus;
  disbursed_at?: string;
}

export interface SavingsRepayment {
  id: number;
  loan_id: number;
  amount: number;
  principal_portion: number;
  interest_portion: number;
  payment_method: SavingsPaymentMethod;
  repaid_at: string;
}

export interface SavingsGroupReport {
  group_id: number;
  name: string;
  status: SavingsGroupStatus;
  interest_rate: number;
  cycle_months: number;
  start_date: string;
  end_date?: string | null;
  total_deposited: number;
  total_borrowed: number;
  total_repaid: number;
  total_interest_collected: number;
  total_receivable: number;
  fund_balance: number;
  total_fund_value: number;
  member_shares: SavingsMember[];
  loans: SavingsLoan[];
}
